### ISSUE TITLE
research: evaluate bounded online calibration for pedestrian goal-force prediction

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#8082]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. The current audit classifies the issue as blocked; its prerequisite, evidence, or triage gate is incomplete.

**First action:** Evaluate whether a frozen pedestrian goal estimator can safely adapt a small calibration/residual layer online using only delayed causal observations, without self-labeling destinations, changing the base model, contaminating evaluation, or creating silent performance drift.

### Current bounded plan

1. Inspect the named owners and dependencies before editing; keep the live issue scope and fail-closed evidence boundary intact.
2. Run the named focused validation, record exact output and remaining blockers, and keep fallback, degraded, unavailable, and inconclusive outcomes explicit.

### Files and owners to inspect

- docs/maintainer_values.md
- docs/context/issue_1512_issue_archetypes.md
- docs/context/pedestrian_goal_force_prediction_research.md
- tests/prediction/test_goal_belief_online_adapter.py
- tests/prediction/test_goal_belief_replay_buffer.py
- tests/prediction/test_goal_belief_update_admission.py
- tests/prediction/test_goal_belief_adapter_rollback.py
- tests/prediction/test_goal_belief_drift_detection.py
- tests/benchmark/test_goal_belief_prequential_eval.py
- configs/benchmarks/goal_belief_online_adaptation_stress.yaml
- scripts/dev/run_worktree_shared_venv.sh
- tests/prediction
- tests/benchmark
- robot_sf/prediction
- robot_sf/benchmark
- scripts/dev/pr_ready_check.sh

### Acceptance and validation

- Demonstrate the stated outcome from the live issue with a focused, reproducible check.
- Keep any missing prerequisite or inconclusive result explicit rather than treating it as success.

### Current dependencies and issue-text corrections

- No supplied individual draft existed for this issue in the handoff package; this is a conservative plan derived from the refreshed live body, and the historical issue text is preserved below.
- Live issue references to verify: https://github.com/ll7/robot_sf_ll7/issues/8060; https://github.com/ll7/robot_sf_ll7/issues/8077; https://github.com/ll7/robot_sf_ll7/issues/8079; https://github.com/ll7/robot_sf_ll7/issues/8081; https://github.com/ll7/robot_sf_ll7/issues/8078; https://github.com/ll7/robot_sf_ll7/issues/8058
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `state:blocked`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `none`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

## Archetype Metadata

```yaml
archetype: implementation
evidence_tier: stress
linked_policy:
  - docs/maintainer_values.md
  - docs/context/issue_1512_issue_archetypes.md
  - docs/context/pedestrian_goal_force_prediction_research.md
```

## Parent and dependencies

- Parent programme: #8060
- Hard predecessors: #8077, #8079, #8081
- Requires a frozen selected estimator or explicit rule-baseline outcome from #8081
- Should consume open-loop forecast evaluation from: #8078
- Research/design basis: PR #8058
- Unblocks: optional adaptation arm in later preregistration; it is not required for the static estimator to proceed

## Goal

Evaluate whether a frozen pedestrian goal estimator can safely adapt a small calibration/residual layer online using only delayed causal observations, without self-labeling destinations, changing the base model, contaminating evaluation, or creating silent performance drift.

The adaptation lane must be optional, default-off, shadow-first, and bounded. Its purpose is to correct observation noise, force-model mismatch, preferred-speed priors, or uncertainty scale—not to let the model reinforce its own destination predictions.

## Research question

Can limited online adaptation improve one-step dynamics consistency and uncertainty calibration under distribution shift while preserving held-out goal/route/change performance and rollback safety?

## Prohibited interpretation

This issue does not establish real-world continual learning, safe autonomous deployment, calibrated human intention, or planner improvement. Simulation oracle labels may score adaptation after the fact but cannot be used as online supervision in the actor-admissible arm.

## Allowed online supervision

At time `t+1`, the actor may obtain delayed causal targets derived from the next observed state:

- next tracked position and velocity;
- one-step velocity/position innovation;
- reconstructed force residual under actor-visible state;
- whether predicted uncertainty covered the observed next state;
- observed association/visibility quality;
- optional recognized arrival-zone label only when the public map and complete track make it independently observable.

The online actor must not use:

- true active/final goal;
- assigned route/waypoint;
- oracle desired-force component;
- simulator identity as a feature;
- future states beyond the delayed target;
- its own top candidate as a pseudo-label;
- a track endpoint caused by occlusion as a final-goal label;
- benchmark test outcomes to alter hyperparameters mid-run.

## Adaptable parameters

Freeze the base rule/model estimator. Permit only a declared bounded adapter, such as:

1. scalar/vector bias correction on reconstructed goal force;
2. low-rank residual head with norm constraints;
3. covariance/process-noise scale;
4. categorical temperature for candidate/change heads;
5. preferred-speed prior mean/variance within physical bounds;
6. force-model mismatch scale by public context class.

Do not adapt the full recurrent/graph backbone in the first implementation. Any full-model adaptation requires a separate issue after this bounded lane passes.

## Adaptation state contract

Per actor/session/domain key record:

```text
adapter_schema_version
base_model_or_rule_digest
adapter_type
adapter_state_digest
update_count
accepted_update_count
rejected_update_count
last_update_step
training_window_start/end
replay_buffer_digest
learning_rate/current bounds
shadow_or_active
rollback_checkpoint_digest
drift_state
blockers
```

The adapter state must reset or persist according to an explicit scope:

- per episode;
- per map/domain session;
- global deployment session.

The smoke must compare at least per-episode and persistent-session semantics. Cross-test contamination is prohibited.

## Replay buffer

Use a bounded, provenance-aware replay buffer with:

- causal sample insertion only after delayed targets arrive;
- maximum rows/age/time span;
- explicit map/scenario/track/segment identity;
- reset/change boundary handling;
- no oracle fields;
- exclusion or separate handling of low-confidence association, saturation, partial force, lost tracks, and fallback rows;
- deterministic sampling with configured RNG;
- class/context balancing fitted without peeking at final outcomes;
- inspectable buffer summary and digest.

A sample cannot cross a confirmed identity change or goal-segment reset unless explicitly allowed for the adapted parameter.

## Update admission

Every candidate update runs first in shadow evaluation against a fixed recent validation buffer or synthetic canaries. Activate only if all declared conditions pass:

- finite bounded parameter delta;
- no schema/model mismatch;
- one-step loss improves by minimum amount;
- uncertainty does not collapse below floor;
- held-out canary score does not regress beyond tolerance;
- change/arrival probabilities remain in valid range;
- runtime/update budget is respected;
- enough high-quality samples exist;
- no drift alarm or data-quality blocker is active.

Rejected updates must leave active bytes unchanged and record the reason.

## Rollback

Maintain an atomic last-known-good adapter state. Roll back when:

- rolling proper score/calibration degrades beyond threshold;
- parameter norm/bounds are violated;
- NaN/Inf appears;
- update latency exceeds budget repeatedly;
- drift is too large for the bounded adapter;
- change false-reset proxy or canary worsens;
- buffer integrity/hash validation fails;
- user/session reset policy requires it.

Rollback must be deterministic, testable, and visible in diagnostics. Silent reset to a default adapter is not acceptable.

## Drift detection

Implement transparent distribution-shift diagnostics over actor features and delayed residuals, such as:

- feature mean/covariance shift;
- residual NLL/coverage shift;
- candidate unknown rate;
- tracking/visibility/saturation mix;
- force-component availability;
- population stability index or another simple bounded statistic;
- optional change-point detector over residual calibration.

Drift detection decides whether adaptation is admissible, not what the true goal is. Large unsupported drift should disable adaptation and mark `out_of_scope`, not aggressively update.

## Experimental arms

On frozen sequential data, compare:

```text
A0 static selected estimator
A1 per-episode covariance/temperature calibration
A2 persistent covariance/temperature calibration
A3 bounded force-bias/residual adapter
A4 bounded prior adaptation
A5 combined bounded adapter
A6 oracle-supervised upper bound, evaluator only and never actor-admissible
```

Use identical sequence order and delayed observation availability. No arm may revisit test data for hyperparameter tuning.

## Shift scenarios

At minimum:

- preferred-speed mean shift;
- relaxation-time/force-factor shift;
- observation noise increase/decrease;
- tracking confidence degradation;
- hidden-neighbour/partial-force increase;
- alternate supported pedestrian model;
- density shift;
- unseen map/route topology;
- increased stopping/restart prevalence;
- changed candidate coverage/unknown rate;
- speed-saturation prevalence shift;
- abrupt domain shift and gradual drift;
- return to the original domain to test catastrophic forgetting/rollback.

## Metrics

Use #8077 and report prequentially over time:

- one-step dynamics NLL/error;
- goal-force/direction proper scores using oracle only after prediction/update timing is fixed;
- candidate/route/change/arrival NLL and calibration;
- coverage/sharpness;
- false reset/detection delay;
- unknown rate;
- forecast propagation metrics via #8078 where feasible;
- update acceptance/rejection/rollback counts;
- adaptation latency and memory;
- performance before/during/after shifts;
- regret versus static baseline;
- forgetting when returning to prior domain;
- oracle upper-bound gap.

The scorer must guarantee that a target used for an update is not also scored as if predicted after the update. Use pre-update prediction for each time step.

## Required tests

Suggested files:

- `tests/prediction/test_goal_belief_online_adapter.py`
- `tests/prediction/test_goal_belief_replay_buffer.py`
- `tests/prediction/test_goal_belief_update_admission.py`
- `tests/prediction/test_goal_belief_adapter_rollback.py`
- `tests/prediction/test_goal_belief_drift_detection.py`
- `tests/benchmark/test_goal_belief_prequential_eval.py`

Required cases:

- base model weights remain byte-identical;
- only whitelisted adapter parameters change;
- no update before delayed target exists;
- oracle/future fields rejected;
- top-prediction pseudo-label cannot enter destination loss;
- low-quality/censored/lost samples follow configured exclusion policy;
- reset/change boundaries isolate samples;
- replay buffer bounded and deterministic;
- rejected update leaves active digest unchanged;
- accepted update satisfies norm/bounds and improves canary objective;
- NaN/Inf and regression trigger rollback;
- persistence/reset scope exact across episodes;
- prequential scorer uses pre-update prediction;
- static/default-off path unchanged;
- repeated same-seed stream yields identical adapter/report digest on deterministic backend.

## Config-first stress evaluation

Provide one frozen sequential stress manifest and canonical runner/checker, conceptually:

```bash
uv run python -m robot_sf.benchmark.goal_belief_online_eval run \
  --config configs/benchmarks/goal_belief_online_adaptation_stress.yaml \
  --input-manifest <frozen-sequence-manifest> \
  --output output/benchmarks/goal_belief_online_adaptation_stress

uv run python -m robot_sf.benchmark.goal_belief_online_eval validate \
  --report output/benchmarks/goal_belief_online_adaptation_stress/report.json
```

Bind stream order, base estimator, initial adapter state, buffer policy, update schedule, shift points, and scorer timing.

## Selection and stop rule

Online adaptation may be admitted to a later closed-loop arm only if it:

1. improves or is non-inferior to the static estimator on prequential proper scores under at least one declared shift;
2. does not materially worsen no-shift held-out performance;
3. retains calibration and change/reset bounds;
4. has no oracle/pseudo-label leakage;
5. demonstrates deterministic rejection and rollback;
6. fits update/runtime/memory budget;
7. does not show unacceptable forgetting after return to the original domain.

If these conditions fail, retain static inference. A negative result is expected to be useful.

## Canonical validation procedure

The PR must report exact results for commands equivalent to:

```bash
scripts/dev/run_worktree_shared_venv.sh -- uv run pytest \
  tests/prediction -k "online_adapter or replay_buffer or rollback or drift" \
  tests/benchmark -k "prequential and goal_belief" -q

scripts/dev/run_worktree_shared_venv.sh -- uv run ruff check \
  robot_sf/prediction robot_sf/benchmark tests/prediction tests/benchmark

scripts/dev/run_worktree_shared_venv.sh -- uv run ruff format --check \
  robot_sf/prediction robot_sf/benchmark tests/prediction tests/benchmark

BASE_REF=origin/main scripts/dev/pr_ready_check.sh
```

Run the frozen stress stream twice and verify report/adapter-event digests.

## Evidence outputs

Produce compact, versioned:

- base estimator/checkpoint/calibrator digests;
- adaptation config and initial-state digest;
- frozen stream/shift manifest;
- update/accept/reject/rollback event ledger;
- buffer quality summary;
- prequential metrics over time and by domain;
- runtime/memory report;
- static-versus-adapted decision;
- independent validation receipt;
- claim boundary `bounded simulation adaptation stress only`.

## Acceptance criteria

- [ ] Base model is frozen and only declared adapter parameters may change.
- [ ] Supervision is delayed, causal, actor-observable, and prequentially scored.
- [ ] Destination heads are never trained on their own predictions.
- [ ] Buffer, update admission, drift detection, and rollback are bounded and deterministic.
- [ ] Shift/no-shift/return-to-domain conditions are evaluated.
- [ ] Proper scores, calibration, change quality, runtime, and forgetting are reported.
- [ ] Default-off static behavior remains unchanged.
- [ ] Oracle upper bound remains separate and planner-inadmissible.
- [ ] Failure produces an explicit static-estimator decision rather than unsafe adaptation.

## Falsification and stop rule

Do not admit online adaptation if actor supervision depends on true goals/forces, if pre-update versus post-update scoring is confused, if no-shift performance or calibration regresses outside bounds, if rollback is not exact, or if the adapter only helps the oracle-clean subset. Close as negative/inconclusive and retain the static estimator.

## Risks and rollback

- Confirmation loop: prohibit destination pseudo-labels and freeze base model.
- Catastrophic forgetting: bounded adapters, replay, return-domain test, rollback.
- Adaptation to tracking errors: quality gates and stratified metrics.
- Evaluation contamination: fixed stream order and prequential timing.
- Rollback: default-off switch plus last-known-good adapter digest.

## Downstream propagation

Update #8060 with whether adaptation is admitted, rejected, or remains diagnostic. Later preregistration must list static and adapted estimators as separate arms and may omit adaptation entirely if this issue is negative.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 8082
source_body_sha256: 2cab61d5192a14b275e18e17fa8c1770aa662cbcfc83a1aac3f173a9cddb1c10
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: blocked
next_action: resolve_named_blocker
authority: blocker_owner
execution_mode: blocker
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-06
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
